cylc/flow/unicode_rules.py

```diff
@@ -0,0 +1,128 @@
+"""Module for unicode restrictions"""
+
+import re
+
+
+ENGLISH_REGEX_MAP = {
+    r'\w': 'alphanumeric',
+    r'\-': '-',
+    r'\.': '.',
+    r'\/': '/'
+}
+
+
+def regex_chars_to_text(chars):
+    r"""Return a string representing a regex component.
+
+    Examples:
+        >>> regex_chars_to_text(['a', 'b', 'c'])
+        ['a', 'b', 'c']
+        >>> regex_chars_to_text([r'\-', r'\.', r'\/'])
+        ['-', '.', '/']
+        >>> regex_chars_to_text([r'\w'])
+        ['alphanumeric']
+
+    """
+    return [
+        ENGLISH_REGEX_MAP.get(char, char)
+        for char in chars
+    ]
+
+
+def length(minimum, maximum):
+    """Restrict character length.
+
+    Example:
+        >>> regex, message = length(0, 5)
+        >>> message
+        'must be between 0 and 5 characters long'
+        >>> bool(regex.match('abcde'))
+        True
+        >>> bool(regex.match('abcdef'))
+        False
+
+    """
+    return (
+        re.compile(r'^.{%d,%d}$' % (minimum, maximum)),
+        f'must be between {minimum} and {maximum} characters long'
+    )
+
+
+def allowed_characters(*chars):
+    """Restrict permitted characters.
+
+    Example:
+        >>> regex, message = allowed_characters('a', 'b', 'c')
+        >>> message
+        'can only contain: a, b, c'
+        >>> bool(regex.match('abc'))
+        True
+        >>> bool(regex.match('def'))
+        False
+
+    """
+    return (
+        re.compile(r'^[%s]+$' % ''.join(chars)),
+        f'can only contain: {", ".join(regex_chars_to_text(chars))}'
+    )
+
+
+def not_starts_with(*chars):
+    """Restrict first character.
+
+    Example:
+        >>> regex, message = not_starts_with('a', 'b', 'c')
+        >>> message
+        'can not start with: a, b, c'
+        >>> bool(regex.match('def'))
+        True
+        >>> bool(regex.match('adef'))
+        False
+
+    """
+    return (
+        re.compile(r'^[^%s]' % ''.join(chars)),
+        f'can not start with: {", ".join(regex_chars_to_text(chars))}'
+    )
+
+
+class UnicodeRuleChecker():
+
+    RULES = []
+
+    @classmethod
+    def __init_subclass__(cls):
```

Comment:
Not a review comment, just pointing out something I found cool (so a review complement, I guess)! This method was new to me, so it was nice to learn a new Python feature by figuring out what its purpose was here, which was to add the rules returned from the methods applied inside ``RULES`` into the doctring of any sub-class:
```console
>>> print(SuiteNameValidator.__doc__)
The rules for valid suite names:

* must be between 1 and 254 characters long
* can not start with: ., -
* can only contain: alphanumeric, /, _, +, -, ., @
```
Nice!